File: OpenStreetMap/iOS/DisplayViewController.m
//
//  SecondViewController.m
//  OSMiOS
//
//  Created by Bryce Cogswell on 12/6/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

#import <sys/utsname.h>

#import "AppDelegate.h"
#import "AerialList.h"
#import "AerialListViewController.h"
#import "EditorMapLayer.h"
#import "MapView.h"
#import "MapViewController.h"
#import "MercatorTileLayer.h"
#import "DisplayViewController.h"


static const NSInteger BACKGROUND_SECTION		= 0;
//static const NSInteger INTERACTION_SECTION	= 1;
static const NSInteger OVERLAY_SECTION			= 2;
static const NSInteger CACHE_SECTION			= 3;


@interface CustomBackgroundCell : UITableViewCell
@property IBOutlet UIButton * button;
@end
@implementation CustomBackgroundCell
@end




@implementation DisplayViewController

-(IBAction)gpsSwitchChanged:(id)sender
{
	// need this to take effect immediately in case they exit the app without dismissing this controller, and they want GPS enabled in background
	MapView * mapView = AppDelegate.getAppDelegate.mapView;
	mapView.enableGpxLogging = _gpxLoggingSwitch.on;
}

-(IBAction)toggleObjectFilters:(UISwitch *)sender
{
	EditorMapLayer * editor = AppDelegate.getAppDelegate.mapView.editorLayer;
	editor.enableObjectFilters = sender.on;
}

- (void)viewWillAppear:(BOOL)animated
{
	[super viewWillAppear:animated];

	MapView * mapView = [(AppDelegate *)[[UIApplication sharedApplication] delegate] mapView];

	if ( [self isMovingToParentViewController] ) {
		// becoming visible the first time
		self.navigationController.navigationBarHidden = NO;

		NSIndexPath * indexPath = [NSIndexPath indexPathForRow:mapView.viewState inSection:BACKGROUND_SECTION];
		UITableViewCell * cell = [self.tableView cellForRowAtIndexPath:indexPath];
		cell.accessoryType = UITableViewCellAccessoryCheckmark;

		[self setCustomAerialCellTitle];

		_notesSwitch.on				= (mapView.viewOverlayMask & VIEW_OVERLAY_NOTES) != 0;
		_gpsTraceSwitch.on			= !mapView.gpsTraceLayer.hidden;

		_birdsEyeSwitch.on			= mapView.enableBirdsEye;
		_rotationSwitch.on			= mapView.enableRotation;
		_unnamedRoadSwitch.on		= mapView.enableUnnamedRoadHalo;
		_gpxLoggingSwitch.on		= mapView.enableGpxLogging;
		_turnRestrictionSwitch.on	= mapView.enableTurnRestriction;
		_objectFiltersSwitch.on		= mapView.editorLayer.enableObjectFilters;

	} else {

		// returning from child view
		[self setCustomAerialCellTitle];
	}
}


- (void)applyChanges
{
	MapView * mapView = [AppDelegate getAppDelegate].mapView;

	NSInteger maxRow = [self.tableView numberOfRowsInSection:BACKGROUND_SECTION];
	for ( NSInteger row = 0; row < maxRow; ++row ) {
		NSIndexPath * indexPath = [NSIndexPath indexPathForRow:row inSection:BACKGROUND_SECTION];
		UITableViewCell * cell = [self.tableView cellForRowAtIndexPath:indexPath];
		if ( cell.accessoryType == UITableViewCellAccessoryCheckmark ) {
			mapView.viewState = (MapViewState)row;
			[mapView setAerialTileService:mapView.customAerials.currentAerial];
			break;
		}
	}
	ViewOverlayMask mask = 0;
	mask |= _notesSwitch.on		? VIEW_OVERLAY_NOTES    : 0;
	mask |= _gpsTraceSwitch.on	? VIEW_OVERLAY_GPSTRACE : 0;
	mapView.viewOverlayMask = mask;

	mapView.enableBirdsEye			= _birdsEyeSwitch.on;
	mapView.enableRotation			= _rotationSwitch.on;
	mapView.enableUnnamedRoadHalo	= _unnamedRoadSwitch.on;
	mapView.enableGpxLogging		= _gpxLoggingSwitch.on;
	mapView.enableTurnRestriction	= _turnRestrictionSwitch.on;
}

- (void)viewWillDisappear:(BOOL)animated
{
	[super viewWillDisappear:animated];

	if ( [self isMovingFromParentViewController] ) {
		[self applyChanges];
	}
}

-(void)setCustomAerialCellTitle
{
	AppDelegate * appDelegate = [AppDelegate getAppDelegate];
	AerialList * aerials = appDelegate.mapView.customAerials;
	NSIndexPath * path = [NSIndexPath indexPathForRow:2 inSection:BACKGROUND_SECTION];
	CustomBackgroundCell * cell = [self.tableView cellForRowAtIndexPath:path];
	if ( [cell isKindOfClass:[CustomBackgroundCell class]] ) {
		[cell.button setTitle:aerials.currentAerial.name forState:UIControlStateNormal];
		[cell.button sizeToFit];
	}
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
	UITableViewCell * cell = [tableView cellForRowAtIndexPath:indexPath];

	if ( indexPath.section == BACKGROUND_SECTION ) {

		// change checkmark to follow selection
		NSInteger maxRow = [self.tableView numberOfRowsInSection:indexPath.section];
		for ( NSInteger row = 0; row < maxRow; ++row ) {
			NSIndexPath * tmpPath = [NSIndexPath indexPathForRow:row inSection:indexPath.section];
			UITableViewCell * tmpCell = [tableView cellForRowAtIndexPath:tmpPath];
			tmpCell.accessoryType = UITableViewCellAccessoryNone;
		}
		cell.accessoryType = UITableViewCellAccessoryCheckmark;

	} else if ( indexPath.section == OVERLAY_SECTION ) {

	} else if ( indexPath.section == CACHE_SECTION ) {

	}
	[self.tableView deselectRowAtIndexPath:indexPath animated:YES];

	// automatically dismiss settings when a new background is selected
	if ( indexPath.section == BACKGROUND_SECTION ) {
		[self.navigationController popToRootViewControllerAnimated:YES];
	}
}

- (void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath
{
	if ( indexPath.section == BACKGROUND_SECTION ) {
		UITableViewCell * cell = [tableView cellForRowAtIndexPath:indexPath];
		cell.accessoryType = UITableViewCellAccessoryNone;
	}
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
	if ( [segue.destinationViewController isKindOfClass:[AerialListViewController class]] ) {
		AerialListViewController * aerialList = segue.destinationViewController;
		aerialList.displayViewController = self;
	}
}

@end

File: src/iOS/DisplaySettings/DisplayViewController.swift
//
//  SecondViewController.h
//  Go Map!!
//
//  Created by Bryce Cogswell on 12/6/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

import MessageUI

private let BACKGROUND_SECTION = 0
private let OVERLAY_SECTION = 2
private let CACHE_SECTION = 3

class DisplayViewController: UITableViewController {
	@IBOutlet var rotationSwitch: UISwitch!
	@IBOutlet var notesSwitch: UISwitch!
	@IBOutlet var questsSwitch: UISwitch!
	@IBOutlet var dataOverlaySwitch: UISwitch!
	@IBOutlet var gpxLoggingSwitch: UISwitch!
	@IBOutlet var turnRestrictionSwitch: UISwitch!
	@IBOutlet var objectFiltersSwitch: UISwitch!
	@IBOutlet var addButtonPosition: UIButton!

	@IBAction func chooseAddButtonPosition(_ sender: Any) {
		let alert = UIAlertController(
			title: NSLocalizedString("+ Button Position", comment: "Location of Add Node button on the screen"),
			message: NSLocalizedString(
				"The + button can be positioned on either the left or right side of the screen",
				comment: ""),
			preferredStyle: .alert)
		let left = UIAlertAction(
			title: NSLocalizedString("Left side", comment: "Left-hand side of screen"),
			style: .default,
			handler: { _ in
				AppDelegate.shared.mainView.settings.buttonLayout = .buttonsOnLeft
				self.setButtonLayoutTitle()
			})
		let right = UIAlertAction(
			title: NSLocalizedString("Right side", comment: "Right-hand side of screen"),
			style: .default,
			handler: { _ in
				AppDelegate.shared.mainView.settings.buttonLayout = .buttonsOnRight
				self.setButtonLayoutTitle()
			})
#if targetEnvironment(macCatalyst)
		// buttons appear right-to-left
		alert.addAction(right)
		alert.addAction(left)
#else
		// buttons appear left-to-right
		alert.addAction(left)
		alert.addAction(right)
#endif
		present(alert, animated: true)
	}

	func applyChanges() {
		let mainView = AppDelegate.shared.mainView!
		let mapView = mainView.mapView!

		let maxRow = tableView.numberOfRows(inSection: BACKGROUND_SECTION)
		for row in 0..<maxRow {
			let indexPath = IndexPath(row: row, section: BACKGROUND_SECTION)
			if let cell = tableView.cellForRow(at: indexPath) {
				if cell.accessoryType == .checkmark {
					mapView.viewState = MapViewState(rawValue: cell.tag) ?? MapViewState.EDITORAERIAL
					mainView.setAerialTileServer(AppState.shared.tileServerList.currentServer)
					break
				}
			}
		}

		mapView.viewOverlayMask = [
			notesSwitch.isOn ? .NOTES : [],
			questsSwitch.isOn ? .QUESTS : [],
			dataOverlaySwitch.isOn ? .DATAOVERLAY : []
		]

		mainView.settings.enableRotation = rotationSwitch.isOn
		mainView.settings.displayGpxTracks = gpxLoggingSwitch.isOn
		mainView.mapLayersView.displayDataOverlayLayers = dataOverlaySwitch.isOn
		mainView.settings.enableTurnRestriction = turnRestrictionSwitch.isOn

		mapView.editorLayer.setNeedsLayout()
	}

	@IBAction func gpsSwitchChanged(_ sender: Any) {
		// need this to take effect immediately in case they exit the app without dismissing this controller, and they want GPS enabled in background
		AppDelegate.shared.mainView.settings.displayGpxTracks = gpxLoggingSwitch.isOn
	}

	@IBAction func dataOverlaySwitchChanged(_ sender: Any) {
		// need this to take effect immediately in case they exit the app without dismissing this controller, and they want GPS enabled in background
		let mainView = AppDelegate.shared.mainView!
		mainView.mapLayersView.displayDataOverlayLayers = dataOverlaySwitch.isOn
	}

	@IBAction func toggleObjectFilters(_ sender: UISwitch) {
		AppDelegate.shared.mapView.editorLayer.objectFilters.enableObjectFilters = sender.isOn
	}

	func setButtonLayoutTitle() {
		let onLeft = AppDelegate.shared.mainView.settings.buttonLayout == MainViewButtonLayout.buttonsOnLeft
		let title = onLeft
			? NSLocalizedString("Left", comment: "")
			: NSLocalizedString("Right", comment: "")
		addButtonPosition.setTitle(title, for: .normal)
	}

	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)

		guard
			let mainView = AppDelegate.shared.mainView,
			let mapView = mainView.mapView
		else { return }

		// becoming visible the first time
		navigationController?.isNavigationBarHidden = false

		notesSwitch.isOn = mapView.viewOverlayMask.contains(.NOTES)
		questsSwitch.isOn = mapView.viewOverlayMask.contains(.QUESTS)
		dataOverlaySwitch.isOn = mainView.mapLayersView.displayDataOverlayLayers

		gpxLoggingSwitch.isOn = AppDelegate.shared.mainView.settings.displayGpxTracks
		turnRestrictionSwitch.isOn = AppDelegate.shared.mainView.settings.enableTurnRestriction
		objectFiltersSwitch.isOn = mapView.editorLayer.objectFilters.enableObjectFilters
		rotationSwitch.isOn = AppDelegate.shared.mainView.settings.enableRotation

		setButtonLayoutTitle()
	}

	override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell,
	                        forRowAt indexPath: IndexPath)
	{
		// place a checkmark next to currently selected display
		if indexPath.section == BACKGROUND_SECTION {
			let mapView = AppDelegate.shared.mapView
			if cell.tag == Int(mapView?.viewState.rawValue ?? -1) {
				cell.accessoryType = .checkmark
			} else {
				cell.accessoryType = .none
			}
		}

		// set the name of the aerial provider
		if indexPath.section == BACKGROUND_SECTION, indexPath.row == 2 {
			if let custom = cell as? CustomBackgroundCell {
				let servers = AppState.shared.tileServerList
				custom.button.setTitle(servers.currentServer.name, for: .normal)
				custom.button.sizeToFit()
			}
		}

		// set the name of the basemap provider
		if indexPath.section == BACKGROUND_SECTION, indexPath.row == 3 {
			if let custom = cell as? CustomBackgroundCell {
				let server = AppDelegate.shared.mainView.mapLayersView.basemapServer
				custom.button.setTitle(server.name, for: .normal)
				custom.button.sizeToFit()
			}
		}
	}

	override func viewWillDisappear(_ animated: Bool) {
		super.viewWillDisappear(animated)
		applyChanges()
	}

	@IBAction func onDone(_ sender: Any?) {
		dismiss(animated: true)
	}

	override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		let cell = tableView.cellForRow(at: indexPath)

		if indexPath.section == BACKGROUND_SECTION {
			// change checkmark to follow selection
			let maxRow = self.tableView.numberOfRows(inSection: indexPath.section)
			for row in 0..<maxRow {
				let tmpPath = IndexPath(row: row, section: indexPath.section)
				let tmpCell = tableView.cellForRow(at: tmpPath)
				tmpCell?.accessoryType = .none
			}
			cell?.accessoryType = .checkmark
		} else if indexPath.section == OVERLAY_SECTION {
		} else if indexPath.section == CACHE_SECTION {}
		self.tableView.deselectRow(at: indexPath, animated: true)

		// automatically dismiss settings when a new background is selected
		if indexPath.section == BACKGROUND_SECTION {
			onDone(nil)
		}
	}

	override func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
		if indexPath.section == BACKGROUND_SECTION {
			let cell = tableView.cellForRow(at: indexPath)
			cell?.accessoryType = .none
		}
	}
}

class CustomBackgroundCell: UITableViewCell {
	@IBOutlet var button: UIButton!
}
